### PR TITLE
Add support for clusters of event reg. refs to WaitForEvents and fix race condition in Observables 

### PR DESCRIPTION
**Pull request by spathiwa**
(The race condition appears to be the source of the long-standing intermittent CI failure in Queue tests.)

### CODE REVIEW

**Comment by spathiwa** [comment]
[Closed and recreated this PR to force the CIs to run again, since AppVeyor got a spurious failure in npm install.)

File: source/core/Events.cpp [hunk 1]
```diff
@@ -396,6 +409,42 @@ class EventRegistrationRefNumManager : public RefNumManager {
 
 EventRegistrationRefNumManager EventRegistrationRefNumManager::_s_singleton;
 
+// GetPendingEventInfo -- given a static queue and a set of dynamic event queuse (passed via refnum), return the queue with the earliest event
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: Fix typo

Comment (spathiwa) [reply source/core/Events.cpp]: Fixed.

File: Makefile [hunk 2]
```diff
@@ -24,7 +24,7 @@ testhttpbin:
 	cd test-it && node test.js -j -t httpbin --dots
 
 testjs:
-	cd test-it && node test.js -j --dots
+	cd test-it && node test.js -j
```
Comment (cglzaguilar) [inline Makefile]: Just for consistency, please make the same change to testnative case. If I remember correctly in failure conditions having the dots is not really helpful.

File: source/CommandLine/main.cpp [hunk 3]
```diff
@@ -115,7 +115,7 @@ int VIREO_MAIN(int argc, const char * argv[])
 void Vireo::RunExec() {
     TypeManagerRef tm = gShells._pUserShell;
     TypeManagerScope scope(tm);
-    Int32 state = tm->TheExecutionContext()->ExecuteSlices(400, 10000000);
+    Int32 state = tm->TheExecutionContext()->ExecuteSlices(10000, 4);
```
Comment (cglzaguilar) [inline source/CommandLine/main.cpp]: Please add a comment as to why we decided to change the value. I know is not clear what would be the best value, but having a history of changes with explanations might help in the future.

File: source/core/Events.cpp [hunk 4]
```diff
@@ -45,6 +45,9 @@ UserEventRefNumManager UserEventRefNumManager::_s_singleton;
 typedef Int32 EventQueueID;
 enum { kNotAQueueID };
 
+// Monotonically incremented event number, so we can always tell which event is generated first even if timestamps are the same
+UInt32 EventData::_s_eventSequenceNumber = 0;
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: Are we concerned about this number wrapping around when the value gets to Max number in the UInt32 range?

File: source/core/Events.cpp [hunk 5]
```diff
@@ -396,6 +409,42 @@ class EventRegistrationRefNumManager : public RefNumManager {
...
+            earliestSeq = eventSeq;
+        } else {
+            eventQID = kNotAQueueID;
+        }
+    }
+    while (regIndex < nQueues) {
+        DynamicEventRegInfo *regInfo = NULL;
+        if (dynRegRefs
+            && EventRegistrationRefNumManager::RefNumStorage().GetRefNumData(dynRegRefs[regIndex].GetRefNum(), &regInfo) == kNIError_Success) {
+            if (_qObject[regInfo->_qID].GetPendingEventSequenceNumber(&eventSeq) && (!eventQID || EventTimeSeqCompare(eventSeq, earliestSeq) < 0)) {
```
Comment (cglzaguilar) [inline source/core/Events.cpp]: For long if statements consider using a local boolean named with the intent of the check, basically the name should answer the question, what is the code checking here? and then use the boolean in the if statement. This is helpful for readability as long as it doesn't have a performance impact.

Apply this rule where you consider is useful in other sections of the code.

File: source/io/module_eggShell.js [hunk 6]
```diff
@@ -517,7 +517,7 @@
         // A good starting point for most vireo uses but can be copied and modified as needed
         // If a callback (stdout, stderr) is provided, it will be run asynchronously to completion
         Module.eggShell.executeSlicesUntilClumpsFinished = publicAPI.eggShell.executeSlicesUntilClumpsFinished = function (callback) {
-            var SLICE_SETS_PER_TIME_CHECK = 100000;
+            var SLICE_SETS_PER_TIME_CHECK = 10000;
```
Comment (cglzaguilar) [inline source/io/module_eggShell.js]: Please add a comment as to why we want this value.

File: Makefile (hunk 2 above)
Comment (spathiwa) [reply Makefile]: This was mistakenly submitted.  It was for debugging a CI failure and I forgot to back it out.  Retracted.

Comment (spathiwa) [reply Makefile]: Although I agree the dots are not helpful, Milan really like them though.  If you agree they should be removed I wouldn't be averse to trying to win over Milan.

File: source/core/Events.cpp (hunk 4 above)
Comment (spathiwa) [reply source/core/Events.cpp]: This is actually handled -- the comparison of this value with previous values if done using a helper function EventTimeSeqCompare which handles wrapping.

File: source/core/Events.cpp (hunk 5 above)
Comment (spathiwa) [reply source/core/Events.cpp]: I agree in general, but this is used in dispatching events, so I am actually concerned about performance.  I have added more comments to help readability, though.

File: source/io/module_eggShell.js (hunk 6 above)
Comment (spathiwa) [reply source/io/module_eggShell.js]: Done.

File: source/CommandLine/main.cpp (hunk 3 above)
Comment (spathiwa) [reply source/CommandLine/main.cpp]: Done; added a reference to the other place in JS with a note they should be in sync.

**Comment by rajsite** [comment]
Since the TFS side does not use executeSlicesUntilClumpsFinished directly, is there a block diagram test that can be written to test the behavior of the loop on that side?

**Comment by spathiwa** [comment]
I'd say we leave them tunable for testing while we discern whether the current values are good or still have issues (or find any reason they might need to be different in different contexts), and then turn them to constants instead of parameters.

**Pull request merged**